libs/vkd3d/swapchain.c

```diff
@@ -2571,26 +2576,35 @@ static void *dxgi_vk_swap_chain_wait_worker(void *chain_)
         pthread_mutex_lock(&chain->wait_thread.lock);
         while (!chain->wait_thread.wait_queue_count)
             pthread_cond_wait(&chain->wait_thread.cond, &chain->wait_thread.lock);
-        next_wait_id = chain->wait_thread.wait_queue[0].id;
-        begin_frame_time_ns = chain->wait_thread.wait_queue[0].begin_frame_time_ns;
+        entry = chain->wait_thread.wait_queue[0];
         pthread_mutex_unlock(&chain->wait_thread.lock);
 
         /* Sentinel for swapchain teardown. */
-        if (!next_wait_id)
+        if (!entry.present_count)
             break;
 
-        cookie = vkd3d_queue_timeline_trace_register_present_wait(timeline_trace, next_wait_id);
-        /* In skip wait mode we just need to make sure that we signal latency fences properly. */
-        if (!chain->wait_thread.skip_waits)
+        if (entry.id)
         {
-            /* We don't really care if we observed OUT_OF_DATE or something here. */
-            VK_CALL(vkWaitForPresentKHR(chain->queue->device->vk_device, chain->present.vk_swapchain,
-                    next_wait_id, UINT64_MAX));
+            cookie = vkd3d_queue_timeline_trace_register_present_wait(timeline_trace, entry.id);
+            /* In skip wait mode we just need to make sure that we signal latency fences properly. */
+            if (!chain->wait_thread.skip_waits)
+            {
+                /* We don't really care if we observed OUT_OF_DATE or something here. */
+                VK_CALL(vkWaitForPresentKHR(chain->queue->device->vk_device, chain->present.vk_swapchain,
+                        entry.id, UINT64_MAX));
+            }
+            vkd3d_queue_timeline_trace_complete_present_wait(timeline_trace, cookie);
+
+            if (entry.begin_frame_time_ns)
+                end_frame_time_ns = vkd3d_get_current_time_ns();
+        }
+        else
+        {
+            d3d12_fence_set_event_on_completion(impl_from_ID3D12Fence1(chain->present.frame_latency_fence),
```

Comment:
We could probably just wait on our own blit semaphore here instead of round-tripping through fence waiter threads and submission thread?

Reply:
Looking at this again we should no longer need the old frame latency fence at all and the complete semaphore is guaranteed to get signaled with the correct timeline anyway, so yeah, good point.